crates/ra_db/src/input.rs

```diff
@@ -114,6 +117,52 @@ impl CrateGraph {
     ) -> impl Iterator<Item = &'a Dependency> + 'a {
         self.arena[&crate_id].dependencies.iter()
     }
+    fn dfs_find(&self, target: CrateId, from: CrateId, visited: &mut FxHashSet<CrateId>) -> bool {
+        if visited.contains(&from) {
+            return false;
+        }
+        for dep in self.dependencies(from) {
+            let crate_id = dep.crate_id();
+            if crate_id == target {
+                return true;
+            }
+
+            if self.dfs_find(target, crate_id, visited) {
+                return true;
+            }
+        }
+        visited.insert(from);
```

Comment:
I believe we need to make a node as visited *on entry*, otherwise there's still a possibility of repeated traversals. So, 

`if !visited.insert(from) { return false; }`